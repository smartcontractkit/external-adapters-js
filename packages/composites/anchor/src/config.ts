import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as DefaultConfig } from '@chainlink/types'

export const DEFAULT_ENDPOINT = 'price'
export const DEFAULT_TOKEN_DECIMALS = 18

export interface Config extends DefaultConfig {
  anchorVaultContractAddress: string
}

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix)
  config.rpcUrl = util.getRequiredEnv('RPC_URL', prefix)
  config.defaultEndpoint = DEFAULT_ENDPOINT

  return {
    ...config,
    anchorVaultContractAddress: util.getRequiredEnv('ANCHOR_VAULT_CONTRACT_ADDRESS', prefix),
  }
}
