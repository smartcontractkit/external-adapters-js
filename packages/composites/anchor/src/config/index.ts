import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as DefaultConfig } from '@chainlink/types'

export const DEFAULT_ENDPOINT = 'price'
export const DEFAULT_TOKEN_DECIMALS = 18
export const NAME = 'ANCHOR'

export const DEFAULT_ANCHOR_VAULT_CONTRACT_ADDRESS = '0xA2F987A546D4CD1c607Ee8141276876C26b72Bdf'
export const DEFAULT_TERRA_BLUNA_CONTRACT_ADDRESS = 'terra1mtwph2juhj0rvjz7dy92gvl6xvukaxu8rfv8ts'

export interface Config extends DefaultConfig {
  anchorVaultContractAddress: string
  terraBLunaContractAddress: string
}

export const makeConfig = (prefix?: string): Config => {
  return {
    ...Requester.getDefaultConfig(prefix),
    rpcUrl: util.getRequiredEnv('RPC_URL', prefix),
    defaultEndpoint: util.getEnv('API_ENDPOINT', prefix) || DEFAULT_ENDPOINT,
    anchorVaultContractAddress:
      util.getEnv('ANCHOR_VAULT_CONTRACT_ADDRESS', prefix) || DEFAULT_ANCHOR_VAULT_CONTRACT_ADDRESS,
    terraBLunaContractAddress:
      util.getEnv('TERRA_BLUNA_CONTRACT_ADDRESS', prefix) || DEFAULT_TERRA_BLUNA_CONTRACT_ADDRESS,
  }
}
