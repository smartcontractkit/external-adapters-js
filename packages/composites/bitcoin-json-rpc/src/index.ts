import { makeExecute } from './adapter'
import { expose } from '@chainlink/ea-bootstrap'

const NAME = 'BITCOIN_JSON_RPC'
export = { NAME, makeExecute, ...expose(NAME, makeExecute()) }
