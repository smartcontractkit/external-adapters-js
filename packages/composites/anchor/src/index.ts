import { expose } from '@chainlink/ea-bootstrap'
import { endpointSelector, makeExecute } from './adapter'
import { makeConfig, NAME } from './config'

/**
 * NOTE: This adapter could probably be made more general in the future.  All it is really doing is fetching an intermediary token price
 * using the Token Allocation adapter and handling the price after that.
 */

const adapterContext = { name: NAME }

const { server } = expose(adapterContext, makeExecute(), undefined, endpointSelector)

export { NAME, makeConfig, makeExecute, server }
