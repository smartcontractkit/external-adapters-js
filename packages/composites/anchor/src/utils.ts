import type {
  AdapterContext,
  AdapterData,
  AdapterRequest,
  AdapterResponse,
} from '@chainlink/ea-bootstrap'
import { BigNumber, ethers } from 'ethers'
import { FIXED_POINT_DECIMALS } from './config'
import * as view from '@chainlink/terra-view-function-adapter'
import { AdapterResponseInvalidError } from '@chainlink/ea-bootstrap'

type ResponseSchema = {
  result: {
    answer: string
  }
}

export const getTokenPrice = async (
  input: AdapterRequest,
  context: AdapterContext,
  feedAddress: string,
  feedDecimals = 8,
): Promise<ethers.BigNumber> => {
  const feedResponse = await callViewFunctionEA<ResponseSchema>(
    input,
    context,
    feedAddress,
    'latest_round_data',
  )
  const latestAnswer = feedResponse.data.result.answer
  const result = ethers.utils.parseUnits(latestAnswer, FIXED_POINT_DECIMALS - feedDecimals)
  throwErrorForInvalidResult(input.id, result, `Chainlink Terra feed ${feedAddress}`)
  return result
}

export const convertUSDQuote = async (
  input: AdapterRequest,
  context: AdapterContext,
  usdPrice: ethers.BigNumber,
  targetQuoteFeedAddress: string,
  feedDecimals = 8,
): Promise<ethers.BigNumber> => {
  const targetQuoteUSDRate = await getTokenPrice(
    input,
    context,
    targetQuoteFeedAddress,
    feedDecimals,
  )
  return usdPrice.mul(BigNumber.from(10).pow(FIXED_POINT_DECIMALS)).div(targetQuoteUSDRate)
}

export const callViewFunctionEA = async <ResponseData extends AdapterData = AdapterData>(
  input: AdapterRequest,
  context: AdapterContext,
  address: string,
  query: string | Record<string, unknown>,
): Promise<AdapterResponse<ResponseData>> => {
  const _config = view.makeConfig()
  const _execute = view.makeExecute(_config)
  const viewFunctionAdapterRequest = {
    id: input.id,
    data: {
      address,
      query,
    },
  }
  return (await _execute(viewFunctionAdapterRequest, context)) as AdapterResponse<ResponseData>
}

export const throwErrorForInvalidResult = (
  jobRunID: string,
  value: ethers.BigNumber,
  label: string,
): void => {
  if (value.lte(BigNumber.from('0')))
    throw new AdapterResponseInvalidError({
      jobRunID,
      statusCode: 500,
      message: `Invalid result for ${label}`,
    })
}
