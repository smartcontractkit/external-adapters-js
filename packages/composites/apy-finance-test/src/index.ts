import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { allocations, tvl } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: tvl.name,
  name: 'APY_FINANCE_TEST',
  config,
  endpoints: [allocations, tvl],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
