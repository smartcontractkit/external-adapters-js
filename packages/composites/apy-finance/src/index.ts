import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig } from './config'

const NAME = 'APY-Finance'

export = { NAME, makeConfig, makeExecute, ...expose(NAME, makeExecute()) }
