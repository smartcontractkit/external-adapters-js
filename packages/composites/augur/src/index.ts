import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig } from './config'

const NAME = 'AUGUR'

export = { NAME, makeConfig, makeExecute, ...expose(NAME, makeExecute()) }
