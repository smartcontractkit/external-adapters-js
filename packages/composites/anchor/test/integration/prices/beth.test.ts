import { AdapterRequest, FastifyInstance } from '@chainlink/ea-bootstrap'
import { server as startServer } from '../../../src'
import nock from 'nock'
import request, { SuperTest, Test } from 'supertest'
import { mockErrorFeedResponse, mockSuccessfulTerraEthFeedResp } from '../fixtures'

import { ethers, BigNumber } from 'ethers'
import '@chainlink/terra-view-function-adapter'
import { AddressInfo } from 'net'

const mockBethStEthResult = BigNumber.from(10).pow(18)
const mockStETHETHPrice = BigNumber.from('1035144096528344468')
const mockZero = BigNumber.from('0')

const ERROR_ETH_FEED = 'error-eth-feed-address'
const FAILED_ETH_FEED = 'failed-eth-feed-address'
const ERROR_STETH_ETH_CURVE_ADDRESS = 'error-curve-address'
const FAILED_STETH_ETH_CURVE_ADDRESS = 'failed-curve-address'
const ERROR_ANCHOR_VAULT_ADDRESS = 'error-anchor-vault-address'
const FAILED_ANCHOR_VAULT_ADDRESS = 'failed-anchor-vault-address'

jest.mock('@chainlink/terra-view-function-adapter', () => {
  return {
    ...jest.requireActual('@chainlink/terra-view-function-adapter'),
    makeExecute: jest.fn().mockReturnValue(
      jest.fn().mockImplementation((input: AdapterRequest) => {
        const address = input.data.address
        switch (address) {
          case ERROR_ETH_FEED:
            return mockErrorFeedResponse
          case FAILED_ETH_FEED:
            throw new Error('Call to ETH-USD Terra feed has reverted')
          default:
            return mockSuccessfulTerraEthFeedResp
        }
      }),
    ),
  }
})

jest.mock('ethers', () => {
  const actualModule = jest.requireActual('ethers')
  return {
    ...actualModule,
    ethers: {
      ...actualModule.ethers,
      providers: {
        JsonRpcProvider: function (): ethers.providers.JsonRpcProvider {
          return {} as ethers.providers.JsonRpcProvider
        },
      },
      Contract: function (address: string) {
        return {
          get_rate: jest.fn().mockImplementation(() => {
            switch (address) {
              case ERROR_ANCHOR_VAULT_ADDRESS:
                return mockZero
              case FAILED_ANCHOR_VAULT_ADDRESS:
                throw new Error('get_rate call to Anchor Hub contract reverted')
              default:
                return mockBethStEthResult
            }
          }),
          get_dy: jest.fn().mockImplementation(() => {
            switch (address) {
              case ERROR_STETH_ETH_CURVE_ADDRESS:
                return mockZero
              case FAILED_STETH_ETH_CURVE_ADDRESS:
                throw new Error('Call to Curve pool to fetch stETH/ETH rate reverted')
              default:
                return mockStETHETHPrice
            }
          }),
        }
      },
    },
  }
})

let oldEnv: NodeJS.ProcessEnv
let fastify: FastifyInstance
let req: SuperTest<Test>
const jobID = '1'

describe('price-beth', () => {
  beforeEach(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))
    fastify = await startServer()
    req = request(`localhost:${(fastify.server.address() as AddressInfo).port}`)
    process.env.COLUMBUS_5_LCD_URL = 'fake-columbus-lcd'
    process.env.API_KEY = 'test'
    process.env.ANCHOR_VAULT_CONTRACT_ADDRESS = 'test-address'
    process.env.COINGECKO_ADAPTER_URL = 'http://localhost:5000'
    process.env.ETHEREUM_RPC_URL = 'test-rpc-url'
    process.env.CACHE_ENABLED = 'false'

    if (process.env.RECORD) {
      nock.recorder.rec()
    }
  })

  afterEach((done) => {
    process.env = oldEnv
    fastify.close(done)
  })

  describe('error calls', () => {
    it('should throw an error if the ETH/USD feed returns 0', async () => {
      process.env.ETH_TERRA_FEED_ADDRESS = ERROR_ETH_FEED
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'ETH',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(500)
      expect(response.body).toMatchSnapshot()
    })

    it('should throw an error if the ETH/USD feed reverts', async () => {
      process.env.ETH_TERRA_FEED_ADDRESS = FAILED_ETH_FEED
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'ETH',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(500)
      expect(response.body).toMatchSnapshot()
    })

    it('should throw an error if the stETH/ETH Curve pool is returns 0', async () => {
      process.env.STETH_POOL_CONTRACT_ADDRESS = ERROR_STETH_ETH_CURVE_ADDRESS
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'ETH',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(500)
      expect(response.body).toMatchSnapshot()
    })

    it('should throw an error if the stETH/ETH Curve pool reverts', async () => {
      process.env.STETH_POOL_CONTRACT_ADDRESS = FAILED_STETH_ETH_CURVE_ADDRESS
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'ETH',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(500)
      expect(response.body).toMatchSnapshot()
    })

    it('should throw an error if the Anchor Vault contract returns 0', async () => {
      process.env.ANCHOR_VAULT_CONTRACT_ADDRESS = ERROR_ANCHOR_VAULT_ADDRESS
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'ETH',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(500)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('successful calls', () => {
    it('return success when fetching the USD/BEth price', async () => {
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'USD',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })

    it('returns success when fetching the ETH/BEth price', async () => {
      const data: AdapterRequest = {
        id: jobID,
        data: {
          from: 'BETH',
          to: 'ETH',
        },
      }

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
