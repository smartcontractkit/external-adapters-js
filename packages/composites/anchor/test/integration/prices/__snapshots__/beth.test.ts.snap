// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`price-beth error calls should throw an error if the Anchor Vault contract returns 0 1`] = `
{
  "error": {
    "feedID": "BETH/ETH",
    "message": "Invalid result for stETH/bETH Exchange Rate from Anchor Vault address error-anchor-vault-address",
    "name": "AdapterError",
    "network": "ethereum",
  },
  "jobRunID": "1",
  "status": "errored",
  "statusCode": 500,
}
`;

exports[`price-beth error calls should throw an error if the ETH/USD feed returns 0 1`] = `
{
  "error": {
    "feedID": "BETH/ETH",
    "message": "Invalid result for Chainlink Terra feed error-eth-feed-address",
    "name": "AdapterError",
  },
  "jobRunID": "1",
  "status": "errored",
  "statusCode": 500,
}
`;

exports[`price-beth error calls should throw an error if the ETH/USD feed reverts 1`] = `
{
  "error": {
    "feedID": "BETH/ETH",
    "message": "Call to ETH-USD Terra feed has reverted",
    "name": "Error",
  },
  "jobRunID": "1",
  "status": "errored",
  "statusCode": 500,
}
`;

exports[`price-beth error calls should throw an error if the stETH/ETH Curve pool is returns 0 1`] = `
{
  "error": {
    "feedID": "BETH/ETH",
    "message": "Invalid result for stETH/ETH Exchange Rate from Curve Pool address error-curve-address",
    "name": "AdapterError",
    "network": "ethereum",
  },
  "jobRunID": "1",
  "status": "errored",
  "statusCode": 500,
}
`;

exports[`price-beth error calls should throw an error if the stETH/ETH Curve pool reverts 1`] = `
{
  "error": {
    "feedID": "BETH/ETH",
    "message": "Call to Curve pool to fetch stETH/ETH rate reverted",
    "name": "AdapterError",
    "network": "ethereum",
  },
  "jobRunID": "1",
  "status": "errored",
  "statusCode": 500,
}
`;

exports[`price-beth successful calls return success when fetching the USD/BEth price 1`] = `
{
  "data": {
    "result": "271217959548",
    "statusCode": 200,
  },
  "jobRunID": "1",
  "result": "271217959548",
  "statusCode": 200,
}
`;

exports[`price-beth successful calls returns success when fetching the ETH/BEth price 1`] = `
{
  "data": {
    "result": "103514409",
    "statusCode": 200,
  },
  "jobRunID": "1",
  "result": "103514409",
  "statusCode": 200,
}
`;
